src/pydvl/value/stopping.py

```diff
@@ -0,0 +1,406 @@
+    """A composable callable object to determine whether a computation
+    must stop.
+
+    A ``StoppingCriterion`` takes a :class:`~pydvl.value.result.ValuationResult`
+    and returns a :class:`~pydvl.value.result.Status~. Objects of this type
+    can be composed with the binary operators ``&`` (*and*), and ``|`` (*or*),
+    following the truth tables of :class:`~pydvl.utils.status.Status`. The
+    unary operator ``~`` (*not*) is also supported.
+
+    :param modify_result: If ``True`` the status of the input
+        :class:`~pydvl.value.result.ValuationResult` is modified in place after
+        the call.
+    """
+
+    # A boolean array indicating whether the corresponding element has converged
+    _converged: NDArray[np.bool_]
+
+    def __init__(self, modify_result: bool = True):
+        self.modify_result = modify_result
+        self._converged = np.full(0, False)
+
+    @abc.abstractmethod
+    def _check(self, result: ValuationResult) -> Status:
+        """Check whether the computation should stop."""
+        ...
+
+    @abc.abstractmethod
+    def completion(self) -> float:
+        """Returns a value between 0 and 1 indicating the completion of the
+        computation.
+        """
+        ...
+
+    def converged(self) -> NDArray[np.bool_]:
+        """Returns a boolean array indicating whether the values have converged
+        for each data point.
+
+        Inheriting classes must set the ``_converged`` attribute in their
+        :meth:`_check`.
+        """
+        return self._converged
+
+    @property
+    def name(self):
+        return type(self).__name__
+
+    def __call__(self, result: ValuationResult) -> Status:
+        status = self._check(result)
+        if self.modify_result:  # FIXME: this is not nice
+            result._status = status
+        return status
+
+    def __and__(self, other: "StoppingCriterion") -> "StoppingCriterion":
+        return make_criterion(
+            fun=lambda result: self._check(result) & other._check(result),
+            completion=lambda: min(self.completion(), other.completion()),
+            name=f"Composite StoppingCriterion: {self.name} AND {other.name}",
+        )(modify_result=self.modify_result or other.modify_result)
+
+    def __or__(self, other: "StoppingCriterion") -> "StoppingCriterion":
+        return make_criterion(
+            fun=lambda result: self._check(result) | other._check(result),
+            completion=lambda: max(self.completion(), other.completion()),
+            name=f"Composite StoppingCriterion: {self.name} OR {other.name}",
+        )(modify_result=self.modify_result or other.modify_result)
+
+    def __invert__(self) -> "StoppingCriterion":
+        return make_criterion(
+            fun=lambda result: ~self._check(result),
+            completion=lambda: 1 - self.completion(),
+            name=f"Composite StoppingCriterion: NOT {self.name}",
+        )(modify_result=self.modify_result)
+
+
+def make_criterion(
```

Comment:
I think this is no longer needed, but I am leaving it here for a while, to see if it's useful. I used to do everything with a variation of this function, returning functions wrapped to have binary ops, and the code was much leaner, but I had to start doing weird things to keep state around and decided for objects instead